File: dsa-problems/MathAdvanced/solutions/python/MTH-005-lagrange-interpolation-mod.py
import sys

class Solution:
    def lagrange_interpolation_mod(self, k: int, X: int, MOD: int, points: list[list[int]]) -> int:
        
        def power(base, exp):
            res = 1
            base %= MOD
            while exp > 0:
                if exp % 2 == 1: res = (res * base) % MOD
                base = (base * base) % MOD
                exp //= 2
            return res

        def modInverse(n):
            return power(n, MOD - 2)

        ans = 0
        
        # O(k^2) approach
        for i in range(k):
            xi, yi = points[i]
            
            num = 1
            den = 1
            
            for j in range(k):
                if i == j: continue
                xj = points[j][0]
                
                num = (num * (X - xj)) % MOD
                den = (den * (xi - xj)) % MOD
            
            term = (yi * num) % MOD
            term = (term * modInverse(den)) % MOD
            ans = (ans + term) % MOD
            
        return (ans + MOD) % MOD

def main():
    input = sys.stdin.read
    data = input().split()
    if not data: return
    
    iterator = iter(data)
    try:
        k = int(next(iterator))
        X = int(next(iterator))
        MOD = int(next(iterator))
        
        points = []
        for _ in range(k):
            points.append([int(next(iterator)), int(next(iterator))])
            
        sol = Solution()
        print(sol.lagrange_interpolation_mod(k, X, MOD, points))
    except StopIteration:
        pass

if __name__ == "__main__":
    main()

File: dsa-problems/MathAdvanced/solutions/cpp/MTH-005-lagrange-interpolation-mod.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution {
    long long MOD;

    long long power(long long base, long long exp) {
        long long res = 1;
        base %= MOD;
        while (exp > 0) {
            if (exp % 2 == 1) res = (res * base) % MOD;
            base = (base * base) % MOD;
            exp /= 2;
        }
        return res;
    }

    long long modInverse(long long n) {
        return power(n, MOD - 2);
    }

public:
    long long lagrange_interpolation_mod(int k, long long X, long long mod, vector<pair<long long, long long>>& points) {
        MOD = mod;
        long long ans = 0;

        for (int i = 0; i < k; i++) {
            long long xi = points[i].first;
            long long yi = points[i].second;

            long long num = 1;
            long long den = 1;

            for (int j = 0; j < k; j++) {
                if (i == j) continue;
                long long xj = points[j].first;

                num = (num * (X - xj + MOD)) % MOD;
                den = (den * (xi - xj + MOD)) % MOD;
            }

            long long term = (yi * num) % MOD;
            term = (term * modInverse(den)) % MOD;
            ans = (ans + term) % MOD;
        }

        return ans;
    }
};

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int k;
    long long X, MOD;
    if (!(cin >> k >> X >> MOD)) return 0;

    vector<pair<long long, long long>> points(k);
    for (int i = 0; i < k; i++) {
        cin >> points[i].first >> points[i].second;
    }

    Solution solution;
    cout << solution.lagrange_interpolation_mod(k, X, MOD, points) << "\n";

    return 0;
}
